### PR TITLE
Added extension Points, initial REST implementation and registering Transport Actions for extensions

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
This PR merges the second batch of feature/extensions into main and will then be backported to 2.x.  It adds extension points, initial REST implementation, and registering Transport Actions for extensions.

### Issues Resolved
Issue with links to PR's included in this merge: https://github.com/opensearch-project/OpenSearch/issues/5271
Meta issue for entire feature/extensions merge: https://github.com/opensearch-project/OpenSearch/issues/5263

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ryanbogan** [comment]
@dblock @reta Can you take a look at this when you get a chance?

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add feature flag for extensions ([#5211](https://github.com/opensearch-project/OpenSearch/pull/5211))
 - Added jackson dependency to server ([#5366] (https://github.com/opensearch-project/OpenSearch/pull/5366))
 - Added experimental extensions to main ([#5347](https://github.com/opensearch-project/OpenSearch/pull/5347))
+- Expanded experimental extensions ([#5518](https://github.com/opensearch-project/OpenSearch/pull/5518))
```
Comment (dblock) [inline CHANGELOG.md]: I think we can merge this with the above.

**Review by dblock — APPROVED** [review]
Explain in the PR what it actually brings in in much more level of detail, please. "Second batch" is not very clear :)

Are we going to backport this to 2.x? Should we be adding the CHANGELOG in the 2.x section?

File: CHANGELOG.md (hunk 1 above)
Comment (ryanbogan) [reply CHANGELOG.md]: Ok so I'll just remove that line then

**Comment by ryanbogan** [comment]
> Are we going to backport this to 2.x? Should we be adding the CHANGELOG in the 2.x section?

@dblock What is the difference between the `Unreleased 2.x` section on main branch and the `Added` section on the CHANGELOG file for 2.x branch?

**Comment by saratvemulapalli** [comment]
@reta would love your thoughts as well on this PR.

**Review by reta — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/common/io/stream/NamedWriteableRegistryParseRequest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    /**
+     * @param in StreamInput from which class fields are read from
+     * @throws IllegalArgumentException if the fully qualified class name is invalid and the class object cannot be generated at runtime
+     */
+    public NamedWriteableRegistryParseRequest(StreamInput in) throws IOException {
+        super(in);
+        try {
+            this.categoryClass = Class.forName(in.readString());
```
[the remark this answers is not in the document]
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/common/io/stream/NamedWriteableRegistryParseRequest.java:48]: The implementation in the up-to-date feature/extensions branch uses a cast `(Class<? extends NamedWriteable>)`.  Would this fix the problem?

[the remark this answers is not in the document]
Comment (reta) [reply server/src/main/java/org/opensearch/common/io/stream/NamedWriteableRegistryParseRequest.java:48]: > NamedWriteable

@ryanbogan I don't think it fixes the problem: the class cast happens **after** the class is loaded and static initializes are called.

**Comment by dblock** [comment]
> > Are we going to backport this to 2.x? Should we be adding the CHANGELOG in the 2.x section?
> 
> @dblock What is the difference between the `Unreleased 2.x` section on main branch and the `Added` section on the CHANGELOG file for 2.x branch?

There shouldn't be AFAIK, but now I am confused too @kotwanikunal?

**Comment by ryanbogan** [comment]
@reta I addressed your requested changes by removing the NamedWriteableRegistry extension implementation, which we will design differently in the future, but do not need right now.  The PR is now ready for your review again.

**Comment by saratvemulapalli** [comment]
> @reta I addressed your requested changes by removing the NamedWriteableRegistry extension implementation, which we will design differently in the future, but do not need right now. The PR is now ready for your review again.

@ryanbogan if we think its not the right implementation, lets go open up an issue and also help us understand why this is not right?

File: server/src/main/java/org/opensearch/extensions/ExtensionBooleanResponse.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.opensearch.transport.TransportResponse;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Generic boolean response indicating the status of some previous request sent to the SDK
+ *
+ * @opensearch.internal
+ */
+public class ExtensionBooleanResponse extends TransportResponse {
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionBooleanResponse.java:22]: It is yet another variation if `.opensearch.index.AcknowledgedResponse` ? May be we could use the same pattern?

**Comment by ryanbogan** [comment]
> > @reta I addressed your requested changes by removing the NamedWriteableRegistry extension implementation, which we will design differently in the future, but do not need right now. The PR is now ready for your review again.
> 
> @ryanbogan if we think its not the right implementation, lets go open up an issue and also help us understand why this is not right?

https://github.com/opensearch-project/OpenSearch/issues/5594

**Comment by ryanbogan** [comment]
@reta ready for your review again.  Thanks for all the feedback!

File: server/src/main/java/org/opensearch/extensions/AcknowledgedResponse.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.opensearch.transport.TransportResponse;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Generic boolean response indicating the status of some previous request sent to the SDK
+ *
+ * @opensearch.internal
+ */
+public class AcknowledgedResponse extends TransportResponse {
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/AcknowledgedResponse.java:22]: 👍 thanks @ryanbogan !

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java [hunk 5]
```diff
@@ -243,39 +306,63 @@ public String executor() {
...
             transportService.sendRequest(
-                extensionNode,
+                extension,
                 REQUEST_EXTENSION_ACTION_NAME,
-                new PluginRequest(transportService.getLocalNode(), new ArrayList<DiscoveryExtensionNode>(uninitializedExtensions)),
-                pluginResponseHandler
+                new InitializeExtensionRequest(transportService.getLocalNode(), extension),
+                initializeExtensionResponseHandler
             );
+            // TODO: make asynchronous
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:317]: Sorry, I remember we discussed that before, this `TODO: make asynchronous` for this pull request or there is issue / pull request to follow up?

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:317]: @reta right here: https://github.com/opensearch-project/OpenSearch/issues/5595

**Review by reta — APPROVED** [review]
LGTM, thanks @ryanbogan

**Review by owaiskazi19 — APPROVED** [review]

**Pull request merged**